### PR TITLE
Keymanager APIs - get,post,delete graffiti

### LINKED ISSUE #13188: Keymanager API: `GET`,`POST`,`DELETE` Graffiti
# 💎 Issue

### Background

 Context and background information on the discussion...

### Description

add `GET`,`POST`,`DELETE` Graffiti endpoints to the validator client

https://github.com/ethereum/keymanager-APIs/pull/63

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

https://ethereum.github.io/keymanager-APIs/?urls.primaryName=dev#/Graffiti

this PR introduces the graffiti keymanager API endpoints ( GET,POST,DELETE ). The approach taken here is adding graffiti to the proposer settings to persist. This also means that only the proposer settings will be saved, if a graffiti is passed via flag it will continue even if the keymanager API is called to delete. That being said the proposer settings now take priority and you will be able to see that in the `GetGraffiti` function.

tests

curl -X POST "http://localhost:7500/eth/v1/validator/0x82f4fca8aacccd676beebda99d662d2e346adcd4f89acb142671f30ded77dda64fc760ef8ed6e62b7bf8cb53dc285778/graffiti" -H  "Content-Type: application/json" -d '{"graffiti":"plain text here"}' -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.1cIoNTAywQNDzFT4_f285omZXuzvSnxLozjtDKxqx-E"
curl -X GET "http://localhost:7500/eth/v1/validator/0x82f4fca8aacccd676beebda99d662d2e346adcd4f89acb142671f30ded77dda64fc760ef8ed6e62b7bf8cb53dc285778/graffiti" -H "accept: application/json"   -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.1cIoNTAywQNDzFT4_f285omZXuzvSnxLozjtDKxqx-E"
{"data":{"pubkey":"0x82f4fca8aacccd676beebda99d662d2e346adcd4f89acb142671f30ded77dda64fc760ef8ed6e62b7bf8cb53dc285778","graffiti":"plain text here"}}

**Which issues(s) does this PR fix?**

Fixes #13188

**Other notes for review**

### CODE REVIEW

File: validator/rpc/handlers_keymanager.go [hunk 1]
```diff
@@ -839,3 +839,87 @@ func (s *Server) DeleteGasLimit(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	rawPubkey, pubkey, ok := shared.HexFromRoute(w, r, "pubkey", fieldparams.BLSPubkeyLength)
+	if !ok {
+		return
+	}
+
+	graffiti, err := s.validatorService.GetGraffiti(ctx, bytesutil.ToBytes48(pubkey))
+	if err != nil {
+		httputil.HandleError(w, errors.Wrap(err, "graffit was not set").Error(), http.StatusNotFound)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:858]: Typo
```suggestion
		httputil.HandleError(w, errors.Wrap(err, "graffiti was not set").Error(), http.StatusNotFound)
```

File: validator/client/propose.go [hunk 2]
```diff
@@ -429,3 +446,34 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
...
+				return v.SetProposerSettings(ctx, ps) // save the proposer settings
+			}
+			return fmt.Errorf("attempted to set graffiti but proposer settings are missing for pubkey:%s", hexutil.Encode(pubKey[:]))
+		}
+	}
+	return errors.New("attempted to set graffiti without proposer settings, graffiti will default to flag options")
+}
+
+func (v *validator) DeleteGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte) error {
+	if v.proposerSettings != nil {
```
Comment (rkapka) [inline validator/client/propose.go:467]: same here

File: validator/rpc/handlers_keymanager.go [hunk 3]
```diff
@@ -839,3 +839,87 @@ func (s *Server) DeleteGasLimit(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	if err := s.validatorService.SetGraffiti(ctx, bytesutil.ToBytes48(pubkey), []byte(req.Graffiti)); err != nil {
+		httputil.HandleError(w, err.Error(), http.StatusNotFound)
+		return
+	}
+}
+
+func (s *Server) DeleteGraffiti(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.SetGraffiti")
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:908]:
```suggestion
	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.DeleteGraffiti")
```

File: validator/accounts/testing/mock.go [hunk 4]
```diff
@@ -213,6 +215,26 @@ func (m *Validator) SetProposerSettings(_ context.Context, settings *validatorse
 	return nil
 }
 
+// GetGraffiti for mocking
+func (m *Validator) GetGraffiti(_ context.Context, _ [fieldparams.BLSPubkeyLength]byte) ([]byte, error) {
+	if m.Graffiti == "" {
+		return nil, errors.New("graffiti is missing ")
+	}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/accounts/testing/mock.go:222]: was just for the mock, but i'll think of a better way to represent

File: validator/client/propose.go [hunk 5]
```diff
@@ -429,3 +446,34 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
 
 	return []byte{}, nil
 }
+
+func (v *validator) SetGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte, graffiti []byte) error {
+	if v.proposerSettings != nil {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/propose.go:451]: oops yeah i should do it this way

File: validator/client/propose.go [hunk 6]
```diff
@@ -382,9 +384,24 @@ func signVoluntaryExit(
...
-// Gets the graffiti from cli or file for the validator public key.
-func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte) ([]byte, error) {
-	// When specified, default graffiti from the command line takes the first priority.
+// GetGraffiti gets the graffiti from cli or file for the validator public key.
+func (v *validator) GetGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte) ([]byte, error) {
+	// Check proposer settings first
+	if v.proposerSettings != nil {
+		if v.proposerSettings.ProposeConfig != nil {
+			option, ok := v.proposerSettings.ProposeConfig[pubKey]
+			if ok && option.Graffiti != "" {
```
Comment (nalepae) [inline validator/client/propose.go:393]: Why checking if `option.Graffiti != ""`?
The user may set an empty graffiti for this precise validator on purpose.

File: validator/client/propose.go [hunk 7]
```diff
@@ -382,9 +384,24 @@ func signVoluntaryExit(
...
+				return []byte(option.Graffiti), nil
+			}
+		}
+		if v.proposerSettings.DefaultConfig != nil {
+			if v.proposerSettings.DefaultConfig.Graffiti != "" {
```
Comment (nalepae) [inline validator/client/propose.go:398]: Same comment here, even if I agree it has less sense...

File: validator/client/propose.go [hunk 8]
```diff
@@ -393,7 +410,7 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
 		return nil, errors.New("graffitiStruct can't be nil")
 	}
 
-	// When specified, individual validator specified graffiti takes the second priority.
+	// When specified, individual validator specified graffiti takes the third priority.
```
Comment (nalepae) [inline validator/client/propose.go:413]: In comments, "third", "fourth" and "fifth" priorities are specified, but not any more "second".

File: validator/client/propose.go [hunk 9]
```diff
@@ -429,3 +446,30 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
 
 	return []byte{}, nil
 }
+
+func (v *validator) SetGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte, graffiti []byte) error {
+	if v.proposerSettings == nil || v.proposerSettings.ProposeConfig == nil {
```
Comment (nalepae) [inline validator/client/propose.go:451]: In this case, why not creating a brand new `v.proposerSettings`?

File: validator/client/propose.go [hunk 10]
```diff
@@ -429,3 +446,30 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
...
+func (v *validator) DeleteGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte) error {
+	if v.proposerSettings == nil || v.proposerSettings.ProposeConfig == nil {
+		return errors.New("attempted to delete graffiti without proposer settings, graffiti will default to flag options")
+	}
+	ps := v.proposerSettings.Clone()
+	option, ok := ps.ProposeConfig[pubKey]
+	if !ok || option == nil {
+		return fmt.Errorf("graffiti not found in proposer settings for pubkey:%s", hexutil.Encode(pubKey[:]))
+	}
+	option.Graffiti = ""
```
Comment (nalepae) [inline validator/client/propose.go:473]: In this case and "empty string" graffiti is the same than no graffiti at all.
A user may set an empty graffiti on purpose.

I guess a solution is to set `option.Graffiti` as a pointer to `Graffiti` and test is the pointer is `nil`.

File: validator/rpc/handlers_keymanager.go [hunk 11]
```diff
@@ -839,3 +839,87 @@ func (s *Server) DeleteGasLimit(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	rawPubkey, pubkey, ok := shared.HexFromRoute(w, r, "pubkey", fieldparams.BLSPubkeyLength)
+	if !ok {
+		return
+	}
+
+	graffiti, err := s.validatorService.GetGraffiti(ctx, bytesutil.ToBytes48(pubkey))
+	if err != nil {
+		httputil.HandleError(w, errors.Wrap(err, "graffiti was not set").Error(), http.StatusNotFound)
```
Comment (nalepae) [inline validator/rpc/handlers_keymanager.go:858]: Here the returned error is `graffiti was not set` even if the error is totally different right?

Comment (nalepae) [reply validator/rpc/handlers_keymanager.go:858]: Example, if I query the graffiti for a key non existing in our VC:
```
{
  "message": "graffiti was not set: rpc error: code = NotFound desc = Could not find validator index for public key 0xa17524bcf06d4fbf042e17dae68637368607f4fc9c29c104f49b12d1e0bed93fa3714f35ff05f4e518f789319c9b8f86",
  "code": 404
}
```

Having both `graffiti was not set` and `Could not find validator index for public key` is quite confusing IMO.

File: validator/client/propose.go (hunk 10 above)
Comment (nalepae) [reply validator/client/propose.go:473]: Example, if I use the following file:

```
default_config:
  fee_recipient: 0x1268AD189526AC0b386faF06eFfC46779c340eE6
  graffiti: Default Graffiti
proposer_config:
  0x80f408a41341416f79e3be605c5f1b960169bea5753b9b8181f7466b4fa0b13addb250c9d068af263f820c113536f9f7:
    fee_recipient: 0xef01d47ed04b181a2715D4050208f7d6e381bF42
  0xa17524bcf06d4fbf042e17dae68637368607f4fc9c28c104f49b12d1e0bed93fa3714f35ff05f4e518f789319c9b8f86:
    fee_recipient: 0x801f87A1e14398e2ec9031CaBd585A84D92BC606
  0x84d7fa31925c559ef85531687b190d9299036625a73370153502bf1c086ac7d7e3006c15210c6f9bd4b518c1900e0f45:
    fee_recipient: 0x801f87A1e14398e2ec9031CaBd585A84D92BC606
    graffiti: ""
```

Then querying the graffiti for `0x84...` will return `Default Graffiti`, while `graffiti: ""` has specifically been set by the user.

File: validator/client/propose.go [hunk 12]
```diff
@@ -433,6 +451,49 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
...
+func (v *validator) SetGraffiti(ctx context.Context, pubkey [fieldparams.BLSPubkeyLength]byte, graffiti []byte) error {
+	if graffiti == nil {
+		return nil
+	}
+	var settings *proposer.Settings
+	if v.proposerSettings == nil {
+		settings = &proposer.Settings{}
+	} else {
+		settings = v.proposerSettings.Clone()
+	}
```
Comment (nalepae) [inline validator/client/propose.go:463]:
```suggestion
    settings = &proposer.Settings{}
	if v.proposerSettings != nil {
		settings = v.proposerSettings.Clone()
	}
```

File: validator/client/propose.go [hunk 13]
```diff
@@ -433,6 +451,49 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
...
+	if settings.ProposeConfig == nil {
+		settings.ProposeConfig = map[[48]byte]*proposer.Option{pubkey: {GraffitiConfig: &proposer.GraffitiConfig{Graffiti: string(graffiti)}}}
+		return v.SetProposerSettings(ctx, settings)
+	}
+	ps := settings.Clone()
+	var option *proposer.Option
```
Comment (nalepae) [inline validator/client/propose.go:469]: This line can be removed.

File: validator/client/propose.go [hunk 14]
```diff
@@ -433,6 +451,49 @@ func (v *validator) getGraffiti(ctx context.Context, pubKey [fieldparams.BLSPubk
...
+	option, ok := ps.ProposeConfig[pubkey]
+	if !ok || option == nil {
+		ps.ProposeConfig[pubkey] = &proposer.Option{GraffitiConfig: &proposer.GraffitiConfig{
+			Graffiti: string(graffiti),
+		}}
+	} else {
+		option.GraffitiConfig = &proposer.GraffitiConfig{
+			Graffiti: string(graffiti),
+		}
+		ps.ProposeConfig[pubkey] = option
```
Comment (nalepae) [inline validator/client/propose.go:479]: Removing this line does not break any test.

File: validator/node/node.go [hunk 15]
```diff
@@ -264,6 +264,7 @@ func (c *ValidatorClient) initializeFromCLI(cliCtx *cli.Context, router *mux.Rou
 		if isWeb3SignerURLFlagSet {
 			c.wallet = wallet.NewWalletForWeb3Signer()
 		} else {
+			fmt.Println("initializeFromCLI asking for wallet")
```
Comment (nalepae) [inline validator/node/node.go:267]: Why a `fmt.Println`?

Comment (james-prysm) [reply validator/node/node.go:267]: that's a mistake from debugging , should remove

**Review by nalepae — DISMISSED** [review]

**Review by rkapka — APPROVED** [review]

**Pull request merged**